### PR TITLE
DAOS-18674 control: Route check engine upcalls/downcalls

### PR DESCRIPTION
**Pull request by kjacque**
This patch moves responsibility for correctly routing check engine reports and repair requests to the control plane, and away from the check leader. This should be transparent to the end user, but lays groundwork for better managing checker activities from the control plane.

- Allow check engines to send CheckReports directly to their
  local control plane via dRPC.
- Forward the CheckReports to the management service over
  gRPC for storage.
- Add control service endpoint for CheckEngineRepair. This
  endpoint sends a repair dRPC to the requested rank.
- Update SystemCheckRepair to call CheckEngineRepair on the
  control plane of the rank where the finding originated.

Features: recovery

### CODE REVIEW

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/1/execution/node/1385/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/1/execution/node/1354/log

**Comment by daosbuild3** [comment]
Test stage NLT completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-17747/2/testReport/

**Comment by daosbuild3** [comment]
Test stage Unit Test completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-17747/2/testReport/

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/2/execution/node/1432/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/2/execution/node/1391/log

File: src/control/server/mgmt_drpc_checker.go [hunk 1]
```diff
@@ -190,43 +195,52 @@ func (mod *srvModule) handleCheckerDeregisterPool(parent context.Context, reqb [
...
+		drpcResp = &sharedpb.CheckReportResp{
+			Status: chkReportErrToDaosStatus(err).Int32(),
 		}
+	} else {
+		drpcResp = &resp.CheckReportResp
 	}
 
-	finding := checker.AnnotateFinding(checker.NewFinding(req.Report))
-	mod.log.Debugf("annotated finding: %+v", finding)
-	if err := mod.checkerDB.AddOrUpdateCheckerFinding(finding); err != nil {
```
Comment (mjmac) [inline src/control/server/mgmt_drpc_checker.go:225]: Looks like `checkerDB` isn't used anymore after this PR -- might be worth nuking it either as part of this PR or a quick follow up to avoid leaving cruft around.

**Review by mjmac — APPROVED** [review]
Looks really good. Just a minor note about cleaning up dead code.

File: src/control/server/mgmt_drpc_checker.go (hunk 1 above)
Comment (kjacque) [reply src/control/server/mgmt_drpc_checker.go:225]: Ah, good catch. This PR will be merging into @Nasf-Fan's, so I can do a follow-on there or here.

File: src/control/lib/control/check_test.go [hunk 2]
```diff
@@ -217,3 +222,212 @@ func TestControl_SystemCheckQuery_ReportsSorted(t *testing.T) {
 		t.Fatalf("reports not sorted (-want +got):\n%s", diff)
 	}
 }
+
+func TestControl_SystemCheckReportRegister(t *testing.T) {
```
Comment (kjacque) [inline src/control/lib/control/check_test.go]: Oops, looks like I forgot to rename this test when I renamed the actual method.

File: src/control/server/mgmt_drpc_checker.go (hunk 1 above)
Comment (kjacque) [reply src/control/server/mgmt_drpc_checker.go:225]: Fixed in latest push since I found something else out of place.

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/4/execution/node/1410/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/5/execution/node/429/log

**Comment by kjacque** [comment]
Had to rebase on parent patch and resolve conflicts. I left the commits basically the same, the conflict was a change in the copyright format.

File: src/control/lib/control/check.go [hunk 3]
```diff
@@ -752,3 +754,109 @@ func SystemCheckRepair(ctx context.Context, rpcClient UnaryInvoker, req *SystemC
...
+// SystemCheckEngineReportReq contains parameters to be passed to SystemCheckEngineReport.
+type SystemCheckEngineReportReq struct {
+	unaryRequest
+	msRequest
+
+	sharedpb.CheckReportReq
+}
+
+// SystemCheckEngineReportResp contains the response from the SystemCheckEngineReport RPC.
+type SystemCheckEngineReportResp struct {
```
Comment (tanabarr) [inline src/control/lib/control/check.go:831]: what's the benefit of aliasing CheckReportResp rather than using directly?

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/6/execution/node/1353/log

**Comment by Nasf-Fan** [comment]
> Test stage Functional Hardware Medium MD on SSD completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/6/execution/node/1353/log

`test_two_pools_corrupted` failed as following:

`2026-04-04 00:53:25,446 test             L0471 INFO | ==> Step 10: Wait for checker to detect inconsistent container label for "dmg check start pool_1 pool_1." [elapsed since last step: 1.42s]
2026-04-04 00:53:25,446 command_utils_ba L0203 DEBUG| Updated param pool => None
2026-04-04 00:53:25,446 general_utils    L0151 INFO | Command environment vars:
  {}
2026-04-04 00:53:25,447 process          L0604 INFO | Running '/usr/bin/dmg -o /var/tmp/daos_testing/configs/daos_control.yml -d -j check query'
2026-04-04 00:53:25,681 process          L0416 DEBUG| [stderr] DEBUG 2026/04/04 00:53:25.681163 main.go:228: debug output enabled
...
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]       "84d301a4-ae0d-4d31-a237-bd97e7191255": {
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "uuid": "84d301a4-ae0d-4d31-a237-bd97e7191255",
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "label": "",
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "status": "CPS_CHECKING",
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "phase": "CSP_POOL_MBS",
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "rank_count": 2,
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "start_time": "2026-04-04T00:53:24.000+00:00",
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "remaining": 9,
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]         "elapsed": 0
2026-04-04 00:53:25,722 process          L0416 DEBUG| [stdout]       },
...
2026-04-04 00:53:25,723 process          L0416 DEBUG| [stdout]     "reports": [
2026-04-04 00:53:25,723 process          L0416 DEBUG| [stdout]       {
2026-04-04 00:53:25,723 process          L0416 DEBUG| [stdout]         "seq": 1331178504193,
2026-04-04 00:53:25,723 process          L0416 DEBUG| [stdout]         "class": 12,
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "action": 7,
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "rank": 1,
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "pool_uuid": "79fd2019-64b0-4436-9dc7-50c110886e8c",
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "pool_label": "TestPool_3",
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "cont_uuid": "5403a7ff-11a0-4069-9ca7-39d2be77d415",
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "cont_label": "TestContainer_1",
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "timestamp": "Sat Apr  4 00:51:17 2026",
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "msg": "Check engine detects inconsistent container label: new-label (CS) vs TestContainer_1 (property).",
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         "act_msgs": [
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]           "Update the CS label to use the container property value for 5403a7ff-11a0-4069-9ca7-39d2be77d415"
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]         ]
2026-04-04 00:53:25,724 process          L0416 DEBUG| [stdout]       }
...
2026-04-04 00:53:26,782 process          L0604 INFO | Running '/usr/bin/dmg -o /var/tmp/daos_testing/configs/daos_control.yml -d -j check repair 1331178504193 2'
2026-04-04 00:53:27,016 process          L0416 DEBUG| [stderr] DEBUG 2026/04/04 00:53:27.016338 main.go:228: debug output enabled
2026-04-04 00:53:27,016 process          L0416 DEBUG| [stderr] DEBUG 2026/04/04 00:53:27.016840 main.go:260: control config loaded from /var/tmp/daos_testing/configs/daos_control.yml
2026-04-04 00:53:27,022 process          L0416 DEBUG| [stderr] DEBUG 2026/04/04 00:53:27.022216 check.go:587: DAOS system check query request: *mgmt.CheckQueryReq (sys:"daos_server-2.9.100"  seqs:1331178504193)
2026-04-04 00:53:27,022 process          L0416 DEBUG| [stderr] DEBUG 2026/04/04 00:53:27.022682 rpc.go:278: request hosts: [hdr-132:10001]
2026-04-04 00:53:27,049 process          L0416 DEBUG| [stderr] DEBUG 2026/04/04 00:53:27.049216 response.go:179: hdr-132:10001: *mgmt.CheckQueryResp (reports:{seq:1331178504193  class:CIC_CONT_BAD_LABEL  action:CIA_TRUST_TARGET  rank:1  pool_uuid:"79fd2019-64b0-4436-9dc7-50c110886e8c"  pool_label:"TestPool_3"  cont_uuid:"5403a7ff-11a0-4069-9ca7-39d2be77d415"  cont_label:"TestContainer_1"  timestamp:"Sat Apr  4 00:51:17 2026"  msg:"Check engine detects inconsistent container label: new-label (CS) vs TestContainer_1 (property)."  act_msgs:"Update the CS label to use the container property value for 5403a7ff-11a0-4069-9ca7-39d2be77d415"})
2026-04-04 00:53:27,050 process          L0416 DEBUG| [stdout] {
2026-04-04 00:53:27,050 process          L0416 DEBUG| [stderr] ERROR: dmg: finding 0x135f06c0001 is already resolved: TRUST_TARGET: Update the CS label to use the container property value for 5403a7ff-11a0-4069-9ca7-39d2be77d415
2026-04-04 00:53:27,050 process          L0416 DEBUG| [stdout]   "response": null,
2026-04-04 00:53:27,050 process          L0416 DEBUG| [stdout]   "error": "finding 0x135f06c0001 is already resolved: TRUST_TARGET: Update the CS label to use the container property value for 5403a7ff-11a0-4069-9ca7-39d2be77d415",
2026-04-04 00:53:27,051 process          L0416 DEBUG| [stdout]   "status": -1025
2026-04-04 00:53:27,051 process          L0416 DEBUG| [stdout] }
2026-04-04 00:53:27,052 process          L0685 INFO | Command '/usr/bin/dmg -o /var/tmp/daos_testing/configs/daos_control.yml -d -j check repair 1331178504193 2' finished with 1 after 0.2686338424682617s
2026-04-04 00:53:27,079 general_utils    L0176 INFO | Error occurred running '/usr/bin/dmg -o /var/tmp/daos_testing/configs/daos_control.yml -d -j check repair 1331178504193 2': Command '/usr/bin/dmg -o /var/tmp/daos_testing/configs/daos_control.yml -d -j check repair 1331178504193 2' failed.
...
`

Please note that the CHK report `1331178504193` belong to former test case. But it was returned by current CHK query. That misguided the test logic.

@kjacque , should we remove such staled CHK report (with seq `1331178504193`)? since related pool has already been destroyed.

**Comment by kjacque** [comment]
> Please note that the CHK report `1331178504193` belong to former test case. But it was returned by current CHK query. That misguided the test logic.
> 
> should we remove such staled CHK report (with seq `1331178504193`)? since related pool has already been destroyed.

Agreed, reports for any destroyed pools should be removed. Let me look deeper into this today. I did a quick Jira search and only saw a couple open issues, so not sure if this is introduced by this PR.

**Comment by kjacque** [comment]
@Nasf-Fan This is an interesting corner case. I think it didn't get cleared because the test only starts the checker on specific pools, so the control plane left alone old findings.

Right now we don't do anything to the finding DB if a pool (or container) is destroyed. Those operations only happen outside checker mode. I wonder if we generally should clear old findings when checker mode is disabled. Is there a reason we'd want to keep old findings?

Edit: Found another ticket that seems like a similar phenomenon, looks like the solution at that point was to update the test. https://daosio.atlassian.net/browse/DAOS-18481

**Comment by Nasf-Fan** [comment]
> @Nasf-Fan This is an interesting corner case. I think it didn't get cleared because the test only starts the checker on specific pools, so the control plane left alone old findings.
> 
> Right now we don't do anything to the finding DB if a pool (or container) is destroyed. Those operations only happen outside checker mode. I wonder if we generally should clear old findings when checker mode is disabled. Is there a reason we'd want to keep old findings?
> 
> Edit: Found another ticket that seems like a similar phenomenon, looks like the solution at that point was to update the test. https://daosio.atlassian.net/browse/DAOS-18481

I do not think we should operate CHK findings DB when destroy the pool/container under regular mode, instead, all related trouble should and can be resolved under check mode. For example, we can introduce new parameter for `dmg check query` command, such as `-f|--full`. If it is specified, then all the history findings (belong to the required pools if specified) will be shown; otherwise (by default), only the findings belong to current check instance (or the latest check instance if current instance has completed) will be shown. On the other hand, we can add new CHK `dmg` command to allow the user to discard specified (stale) findings from DB if he/she want. But of course, these can be done in another dedicated patch.

How do you think? @kjacque

File: src/control/server/ctl_check.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/pkg/errors"
+	"google.golang.org/protobuf/proto"
+
+	ctlpb "github.com/daos-stack/daos/src/control/common/proto/ctl"
+	mgmtpb "github.com/daos-stack/daos/src/control/common/proto/mgmt"
+	"github.com/daos-stack/daos/src/control/lib/daos"
+	"github.com/daos-stack/daos/src/control/lib/ranklist"
+)
+
+func (svc *ControlService) findRankIndex(rank ranklist.Rank) (uint32, error) {
```
Comment (tanabarr) [inline src/control/server/ctl_check.go]: I feel like this is done in multiple places, maybe we should centralise?

**Review by mjmac — APPROVED** [review]

**Comment by kjacque** [comment]
> I do not think we should operate CHK findings DB when destroy the pool/container under regular mode, instead, all related trouble should and can be resolved under check mode. For example, we can introduce new parameter for `dmg check query` command, such as `-f|--full`. If it is specified, then all the history findings (belong to the required pools if specified) will be shown; otherwise (by default), only the findings belong to current check instance (or the latest check instance if current instance has completed) will be shown. On the other hand, we can add new CHK `dmg` command to allow the user to discard specified (stale) findings from DB if he/she want. But of course, these can be done in another dedicated patch.
> 
> How do you think? @kjacque

I agree we should probably not be touching the findings in normal mode. If we want to track the findings according to check instance, we will probably need some kind of sequence number to label the check instance. Not sure if it would need to be tracked at the engine level--maybe the control plane could track this on its own.

Earlier today in triage we saw [a ticket for the same issue seen on master](https://daosio.atlassian.net/browse/DAOS-18773), so we can continue the discussion there. Since it seems this patch is otherwise OK, and @mjmac approved, shall we merge it into your branch? @Nasf-Fan

**Comment by Nasf-Fan** [comment]
> > I do not think we should operate CHK findings DB when destroy the pool/container under regular mode, instead, all related trouble should and can be resolved under check mode. For example, we can introduce new parameter for `dmg check query` command, such as `-f|--full`. If it is specified, then all the history findings (belong to the required pools if specified) will be shown; otherwise (by default), only the findings belong to current check instance (or the latest check instance if current instance has completed) will be shown. On the other hand, we can add new CHK `dmg` command to allow the user to discard specified (stale) findings from DB if he/she want. But of course, these can be done in another dedicated patch.
> > How do you think? @kjacque
> 
> I agree we should probably not be touching the findings in normal mode. If we want to track the findings according to check instance, we will probably need some kind of sequence number to label the check instance. Not sure if it would need to be tracked at the engine level--maybe the control plane could track this on its own.
> 
> Earlier today in triage we saw [a ticket for the same issue seen on master](https://daosio.atlassian.net/browse/DAOS-18773), so we can continue the discussion there. Since it seems this patch is otherwise OK, and @mjmac approved, shall we merge it into your branch? @Nasf-Fan

Yes, we can handle `dmg check query` related issues in DAOS-18773.

I am not familiar with control plane implementation. @tanabarr , would you please to help another review for the huge patch before merging the patch? Thanks!

File: src/control/server/mgmt_check.go [hunk 5]
```diff
@@ -600,22 +612,77 @@ func (svc *mgmtSvc) SystemCheckRepair(ctx context.Context, req *mgmtpb.CheckActR
...
+	if err := svc.checkLeaderRequest(wrapCheckerReq(req)); err != nil {
+		return nil, err
+	}
+
+	if err := svc.verifyCheckerReady(); err != nil {
+		return nil, err
+	}
+
+	if req.Report.PoolLabel == "" && req.Report.PoolUuid != "" {
+		svc.log.Debug("looking up pool UUID")
```
Comment (tanabarr) [inline src/control/server/mgmt_check.go]: is this necessary?

File: src/proto/ctl/check.proto [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+//
+
+// This file defines check engine-related messages.
+
+syntax = "proto3";
+package ctl;
+
+option  go_package = "github.com/daos-stack/daos/src/control/common/proto/ctl";
+
+import "mgmt/check.proto";
```
Comment (tanabarr) [inline src/proto/ctl/check.proto:14]: should the base check types be declared in shared if they need to be used in both ctl and mgmt packages?

**Review by tanabarr — APPROVED** [review]
src/engine/drpc_ras.c formatting-only changes should be reverted
findRankIndex could be consolidated with other similar code

File: src/control/lib/control/check.go (hunk 3 above)
Comment (kjacque) [reply src/control/lib/control/check.go:831]: Consistency in the naming mostly, but it also gives us some flexibility with implementation in the future without altering the API types. With a lot of the checker stuff we just wrap the protobuf structures, rather than copying the structures (which is a more typical approach elsewhere in the control lib).

File: src/control/server/ctl_check.go (hunk 4 above)
Comment (kjacque) [reply src/control/server/ctl_check.go]: Actually looks like there's a helper function on the harness `FilterInstancesByRankSet`... I'll use that instead. Thanks.

File: src/control/server/mgmt_check.go (hunk 5 above)
Comment (kjacque) [reply src/control/server/mgmt_check.go]: Good point, I found it helpful in development but I can make it more informative and downgrade to trace.

File: src/proto/ctl/check.proto (hunk 6 above)
Comment (kjacque) [reply src/proto/ctl/check.proto:14]: I thought this initially too, but it ended up being a bit painful and ugly at the engine level, figuring out which files to export where. `CheckActReq` is needed in `src/mgmt`, while other types in the `shared` namespace wind up in `src/engine`... Seemed less confusing to leave the cross-module import. If you feel strongly I can change it.

File: src/control/server/ctl_check.go (hunk 4 above)
Comment (kjacque) [reply src/control/server/ctl_check.go]: Fixed in latest push.

File: src/control/server/mgmt_check.go (hunk 5 above)
Comment (kjacque) [reply src/control/server/mgmt_check.go]: Fixed

**Review by tanabarr — APPROVED** [review]

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/8/execution/node/417/log

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/8/execution/node/458/log

**Comment by Nasf-Fan** [comment]
> Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/8/execution/node/458/log

Before the latest two commits, we can pass [cat_recov_core.py], but failed now. Need further investigation before merging.

**Comment by Nasf-Fan** [comment]
> > Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/8/execution/node/458/log
> 
> Before the latest two commits, we can pass [cat_recov_core.py], but failed now. Need further investigation before merging.



> > Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/8/execution/node/458/log
> 
> Before the latest two commits, we can pass [cat_recov_core.py], but failed now. Need further investigation before merging.

It is verified that the CR test failure were caused by test logic instead of the patch.

But I found some `dmg check enable` failure during my local CR test, not sure whether it is related with the patch or not, something like the following:


[       OK ] CR20: check engine death during check
[ RUN      ] CR21: check engine rejoins check instance successfully
CR21: check engine rejoins check instance successfully
CR: creating pool ...
setup: creating pool, SCM size=0 GB, NVMe size=0 GB
setup: created pool a0a829c1-102e-4371-b5ad-08c2887248ec
CR: getting label for pool a0a829c1
CR: connecting pool a0a829c1
CR: injecting fault CIC_POOL_NONEXIST_ON_MS for pool a0a829c1
CR: stopping system with non-force ...
CR: enable check mode
2026/04/13 13:15:03.780010 fedora-linux-40 DAOS[187138/187138/0] daos ERR  src/common/tests_dmg_helpers.c:391 daos_dmg_json_pipe() dmg error: client: code = 510 description = "the *control.SystemCheckEnableReq request timed out after 5m0s"
2026/04/13 13:15:03.780021 fedora-linux-40 DAOS[187138/187138/0] daos ERR  src/common/tests_dmg_helpers.c:1856 dmg_check_switch() dmg check switch to enable failed: -1025
Failure assert_rc_equal src/tests/suite/daos_cr.c:2993 DER_MISC(-1025) != DER_SUCCESS(0)
[  ERROR   ] --- "DER_MISC" != "DER_SUCCESS"


@kjacque , will your patch potentially affect `dmg check enable/disable` behavior or not? Thanks!

**Comment by Nasf-Fan** [comment]
@kjacque , can we make `dmg check query` output to contains the following information: check leader and report sponsor. For example:

  "response": {
    "start_time": "2026-04-12T00:41:30.000+00:00",
    "status": "COMPLETED",
    "scan_phase": "DONE",
    **`"leader": 3,`**
    "pools": {
      "7f6a886c-63bd-47f7-8aa4-acba6a4fba9b": {
        "uuid": "7f6a886c-63bd-47f7-8aa4-acba6a4fba9b",
        "label": "",
        "status": "CPS_CHECKED",
        "phase": "CSP_DONE",
        "rank_count": 8,
        "start_time": "2026-04-12T00:41:30.000+00:00",
        "remaining": 0,
        "elapsed": 9
      }
    },
    "reports": [
      {
        "seq": 4723950485505,
        "class": 7,
        "action": 5,
        **`"rank": 4,`**
        "pool_uuid": "7f6a886c-63bd-47f7-8aa4-acba6a4fba9b",
        "pool_label": "test_fPZkbx-fault",
        "timestamp": "Sun Apr 12 00:41:40 2026",
        "msg": "Check engine detects corrupted pool label: test_fPZkbx-fault (MS) vs test_fPZkbx (PS).",
        "act_msgs": [
          "Reset the pool

That will be helpful. Thanks!

**Review by daltonbohning — COMMENTED** [review]
ftest LGTM

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/9/execution/node/1315/log

**Comment by kjacque** [comment]
@Nasf-Fan If hardware testing passes, let's merge this into your branch.

File: src/control/lib/control/check.go [hunk 7]
```diff
@@ -754,6 +755,57 @@ func SystemCheckRepair(ctx context.Context, rpcClient UnaryInvoker, req *SystemC
...
+func CheckEngineRepair(ctx context.Context, rpcClient UnaryInvoker, req *CheckEngineRepairReq) (*CheckEngineRepairResp, error) {
+	if req == nil {
+		return nil, errors.Errorf("nil %T", req)
+	}
+
+	if req.Req == nil {
+		return nil, errors.Errorf("no action request included in %T", req)
+	}
+
+	if len(req.HostList) != 1 {
```
Comment (tanabarr) [inline src/control/lib/control/check.go:782]: we should probably generalise this single host requirement in to an embedded struct type at some point

File: src/control/lib/control/check.go [hunk 8]
```diff
@@ -403,6 +403,17 @@ type SystemCheckReport struct {
 	chkpb.CheckReport
 }
 
+func (r *SystemCheckReport) MarshalJSON() ([]byte, error) {
+	type toJSON SystemCheckReport
+	return json.Marshal(struct {
+		Rank uint32 `json:"rank"`
```
Comment (tanabarr) [inline src/control/lib/control/check.go:409]: Nice hack, is it worth a unit test to show the effect when printing?

**Review by tanabarr — APPROVED** [review]
Great work!

File: src/control/lib/control/check.go (hunk 8 above)
Comment (kjacque) [reply src/control/lib/control/check.go:409]: I was wondering the same... I was able to test manually, at least. A unit test wouldn't be too hard, but I'd rather address as a follow-on to avoid CI churn.

File: src/control/lib/control/check.go (hunk 7 above)
Comment (kjacque) [reply src/control/lib/control/check.go:782]: Agreed, we have a few of these.

**Comment by daosbuild3** [comment]
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/10/execution/node/1316/log

**Comment by Nasf-Fan** [comment]
> Test stage Functional Hardware Medium MD on SSD completed with status FAILURE. https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-17747/10/execution/node/1316/log

test_dfuse_daos_build_wb failed for DAOS-18813.
test_two_pools_corrupted failure will be handled via DAOS-18773.
test_snapshot_aggregation failed for some engine level conditional insert timeout, seems not related with the control plane changes.

**Review by Nasf-Fan — APPROVED** [review]
The patch is fine for me. Thanks @kjacque !

**Review by mjmac — APPROVED** [review]
